src/services/WaveformDataUtils.js

```diff
@@ -161,15 +165,55 @@ export default class WaveformDataUtils {
    * @param {Object} peaksInstance - current peaks instance for the waveform
    */
   activateSegment(id, peaksInstance) {
-    const segment = peaksInstance.segments.getSegment(id);
+    const validatedPeaks = this.initialSegmentValidation(id, peaksInstance);
+    const segment = validatedPeaks.segments.getSegment(id);
     segment.update({
       editable: true,
       color: COLOR_PALETTE[2]
     });
     let startTime = segment.startTime;
     // Move play head to the start time of the selected segment
-    peaksInstance.player.seek(startTime);
+    validatedPeaks.player.seek(startTime);
 
+    return validatedPeaks;
+  }
+
+  /**
+   * Adjust segment's end time to depict the valid time range for it to be spanned
+   * At initial load of the editor for all segments with invalid end times,
+   * segment.endTime is set to duration by default
+   * @param {String} id - ID of the segment being edited
+   * @param {Object} peaksInstance - current peaks instance for the waveform
+   */
+  initialSegmentValidation(id, peaksInstance) {
+    const segment = peaksInstance.segments.getSegment(id);
+    const duration = Math.round(peaksInstance.player.getDuration() * 100) / 100;
+    // Segments before and after the current segment
+    const { before, after } = this.findWrapperSegments(segment, peaksInstance);
+
+    // Check for margin of +/- 0.02 milliseconds to be considered
+    let isDuration = time => {
+      return time <= duration + 0.02 && time >= duration - 0.02;
+    };
+    if (
+      before &&
+      segment.startTime < before.endTime &&
+      !isDuration(before.endTime)
+    ) {
+      segment.update({ startTime: before.endTime });
+    } else if (isDuration(segment.endTime)) {
+      const allSegments = this.sortSegments(peaksInstance, 'startTime');
+      segment.update({
+        endTime: allSegments.filter(seg => seg.startTime > segment.startTime)[0]
+          .startTime
+      });
+    } else if (
+      after &&
+      segment.endTime > after.startTime &&
+      !isDuration(after.endTime)
```

Comment:
Does it matter that the after.endTime isn't duration?  It seems like we only care about after.startTime.

Reply:
Yes, we don't need to check for that :+1: